### PR TITLE
promtool: backfill: allow configuring block duration

### PR DESCRIPTION
**Pull request by stevekuznetsov**
When backfilling large amounts of data across long periods of time, it
may in certain circumstances be useful to use a longer block duration to
increase the efficiency and speed of the backfilling process. This patch
adds a flag --block-duration-power to allow a user to choose the power N
where the block duration is 2^(N+1)h.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

---
Fixes https://github.com/prometheus/prometheus/issues/8388
Builds on top of #8917 so as to not create merge conflicts. I do not know the best place to leave the discussion about block durations, so I added it to the helptext. I certainly think it is useful for users who are not familiar with TSDB basics to have that discussion present when they are choosing a block duration.
/assign @roidelapluie @brian-brazil

### CODE REVIEW

**Comment by roidelapluie** [comment]
Thanks.

I prefer to have the duration of the blocks, as it would be more telling for the users. Power is really obscure and it's an implementation detail.

We could in the help list a few examples (2h, 4h, 8h).

If you think it is too much to ask for the user, we could round down the duration provided (with a warning). e.g. if you pass 20h you would have 16h blocks. But I don't think it is necessary for a first step.

Note: the documentation should be on the storage page.

**Comment by stevekuznetsov** [comment]
> Note: the documentation should be on the storage page.

@roidelapluie sorry, would you mind linking where this page is, so I could edit it?

**Comment by stevekuznetsov** [comment]
Moved the conversation about when to use the flag to `docs/storage.md`

**Comment by roidelapluie** [comment]
What if, if we reject, we print to stderr the value below and above? So the user can pick easily?

I am really affraid that users will write "24" or even "10", making gigantic blocks by accident, plus this is adding a lot of complexity (you need to compute the power instead of picking from the few examples we show in the --help)

**Comment by dgl** [comment]
The other way to think about this is this is an API (ok, outside the Prometheus API stability guarantees currently), but by forcing this to be a particular power it means this becomes coupled to the implementation of the block storage.

By using a duration that is validated, it's nicer to users but also does not get in the way of any potential (even if unlikely) changes to the block format.

**Comment by stevekuznetsov** [comment]
@dgl the maintainers made it very clear to me that a) only very specific values are allowed b) if you use other values, things will go very wrong. What is the benefit of planning for an unlikely future where the default block size changes? I want to build a useful tool _today_ that makes it impossible for users to make the wrong decision, in a space where the decision is hard to make and incorrect choices are deleterious.

**Comment by stevekuznetsov** [comment]
@roidelapluie IDK if you and other maintainers feel really strongly I can change it but I honestly believe that giving users an infinite input space, making it *seem* like there's choice, and validating it down to very, very distinct values is simply terrible UX. Compare the two cases:

1. the user has a `--block-duration-power` flag, reads the documentation on how to use it, chooses a power and goes from there
2. the user has a `--block-duration` flag, makes a guess, chooses an invalid duration, has to go read the doc to figure out which values are valid, implicitly chooses a power, goes from there

Why is 2 better?

**Comment by roidelapluie** [comment]
I am quite confident that we should not expose those kind of details to our users.

What we could to is to propose a --max-block-duration=2h option instead, and we take the power of 2 below it.

--max-block-duration=24h would create 16h blocks.

**Comment by stevekuznetsov** [comment]
@roidelapluie that is a good approach as well. Do we expect users to want a specific max value, or should we add something like `--use-largest-blocks` that looks at the input OpenMetrics data and simply chooses the block size that would fit all of the data? Or maybe a `--split-across-blocks=N` and choose some "correct" duration whereby N blocks result?

**Comment by roidelapluie** [comment]
--use-largest-blocks will probably be too big for e.g. 1 year of data. `--split-across-blocks=N` might work but it would also lie to the user which has data for less than e.g. 2 hours.

**Comment by stevekuznetsov** [comment]
@roidelapluie what would be reasonable bounds for the block size? No smaller than 2h and no larger than a week? A month? three months?

**Comment by stevekuznetsov** [comment]
@roidelapluie ok - I've added the flag to choose a number of blocks and added validation so it's not too large or too small. I chose a random value for the largest allowable block, but please take a look and advise me as to the best option here.

**Comment by stevekuznetsov** [comment]
(I think we'll need to do the close/reopen thing here too to get CircleCI to run?)

**Comment by roidelapluie** [comment]
I still don't think this is the correct fix. I stand by my proposal of a max block duration because that is what you pass to the prometheus server. Then you can have the same max block duration on backfill and prometheus. Prometheus does not have the option "keep 10 blocks".

**Comment by stevekuznetsov** [comment]
@roidelapluie you should have said so... your previous comments definitely sounded like you endorsing this approach. With a duration, though, could you please explain to me why it's the better approach to ask the user for a raw duration when almost all possible values are invalid?

**Comment by roidelapluie** [comment]
If passed storage.tsdb.max-block-duration=24h, prometheus would create blocks up to 16h, like in my proposal. I would not error.

**Comment by roidelapluie** [comment]
We ask the user for a max duration, not for an exact duration. That mimics the options of Prometheus, so it is safe for users to use the same value as their Prometheus server.

**Comment by stevekuznetsov** [comment]
Well, ya should have said that! If it's an existing UX used elsewhere, there's nothing to do but copy it here :)

**Comment by stevekuznetsov** [comment]
(to be fair, the code you linked me that you wrote did the validation and sent an error... I would really love to have learned of the Prometheus TSDB flag earlier so we could have saved SO much time talking about all this ...)

**Comment by stevekuznetsov** [comment]
Code now uses identical logic to the `--storage.tsdb.max-block-duration` flag in the main `prometheus` binary.

**Comment by stevekuznetsov** [comment]
@roidelapluie I'm not sure how you got CircleCI to work on my other PR but I think you will need to use that magic again here :|

File: docs/storage.md [hunk 1]
```diff
@@ -157,6 +157,19 @@ promtool tsdb create-blocks-from openmetrics <input file> [<output directory>]
...
+
+- the block duration *must* be a power of two number of hours
+- blocks, when inserted must be aligned to the TSDB
+
+The --max-block-duration flag allows the user to configure some maximum duration of blocks. The backfilling tool will pick a block duration no larger than this that follows the above rules.
+
+While larger blocks may improve the performance of backfilling large datasets, drawbacks exist as well, as many algorithms in the TSDB will operate over the entire block as one monolithic entity. The larger a block is when written to the TSDB, the higher chances are that adverse effects will occur: time-based retention policies must keep the entire block around if even one part of the (potentially large) block is still valid. Conversely, size-based retention policies will remove the entire contents of the block even if the TSDB only goes over the size limit in a minor way.
+
+Therefore, backfilling with few blocks, thereby choosing a larger block duration, must be done with care and is not recommended for any production instances.
```
Comment (codesome) [inline docs/storage.md:171]: Some suggested edits to make it more compact and removed some explanation about block range selection which is not relevant for this section of the doc, especially when it's `max-block-duration` now.

Regarding `"not recommended for any production instances"`, I am not sure, any comments on that @roidelapluie?

```suggestion
#### Longer Block Durations

By default, the promtool will use the default block duration (2h) for the blocks; this behavior is the most generally applicable and correct. However, when backfilling data over a long range of times, it may be advantageous to use a larger value for the block duration to backfill faster and prevent additional compactions by TSDB later.

The `--max-block-duration` flag allows the user to configure a maximum duration of blocks. The backfilling tool will pick a suitable block duration no larger than this.

While larger blocks may improve the performance of backfilling large datasets, drawbacks exist as well. Time-based retention policies must keep the entire block around if even one sample of the (potentially large) block is still within the retention policy. Conversely, size-based retention policies will remove the entire block even if the TSDB only goes over the size limit in a minor way.

Therefore, backfilling with few blocks, thereby choosing a larger block duration, must be done with care and is not recommended for any production instances.
```

**Review by codesome — COMMENTED** [review]
Thanks for the PR! Minor changes required

File: cmd/promtool/backfill.go [hunk 2]
```diff
@@ -65,8 +66,17 @@ func getMinAndMaxTimestamps(p textparse.Parser) (int64, int64, error) {
...
+func createBlocks(input []byte, mint, maxt, maxBlockDuration int64, maxSamplesInAppender int, outputDir string, humanReadable, quiet bool) (returnErr error) {
 	blockDuration := tsdb.DefaultBlockDuration
+	if maxBlockDuration > 0 {
+		ranges := tsdb.ExponentialBlockRanges(tsdb.DefaultBlockDuration, 10, 3)
+		for i, v := range ranges {
+			if v > maxBlockDuration {
+				blockDuration = ranges[i-1]
+				break
+			}
+		}
```
[the remark this answers is not in the document]
Comment (stevekuznetsov) [reply cmd/promtool/backfill.go:78]: FWIW the code for the TSDB flag should be updated to do that too!

**Comment by stevekuznetsov** [comment]
@codesome thanks for the review! Updated

File: cmd/promtool/backfill.go [hunk 3]
```diff
@@ -65,8 +66,19 @@ func getMinAndMaxTimestamps(p textparse.Parser) (int64, int64, error) {
 	return maxt, mint, nil
 }
 
-func createBlocks(input []byte, mint, maxt int64, maxSamplesInAppender int, outputDir string, humanReadable, quiet bool) (returnErr error) {
+func createBlocks(input []byte, mint, maxt, maxBlockDuration int64, maxSamplesInAppender int, outputDir string, humanReadable, quiet bool) (returnErr error) {
 	blockDuration := tsdb.DefaultBlockDuration
+	if maxBlockDuration > 0 && maxBlockDuration > tsdb.DefaultBlockDuration {
+		ranges := tsdb.ExponentialBlockRanges(tsdb.DefaultBlockDuration, 10, 3)
+		idx := len(ranges) - 1 // use largest range if user asked for something enormous
```
Comment (codesome) [inline cmd/promtool/backfill.go:73]: Nit: Fixing my half baked suggestion earlier plus the comment style that we follow (start with caps, end with a `.`)
```suggestion
	if maxBlockDuration > tsdb.DefaultBlockDuration {
		ranges := tsdb.ExponentialBlockRanges(tsdb.DefaultBlockDuration, 10, 3)
		idx := len(ranges) - 1 // Use largest range if user asked for something enormous.
```

**Review by codesome — COMMENTED** [review]
Sorry for not noticing it before: we should be adding a couple of test cases in `backfill_test.go -> TestBackfill` with higher max block duration which does not produce 2h blocks. You can add an  optional variable in `tests` variable for this.

**Comment by stevekuznetsov** [comment]
@codesome  of course - added a test for a longer duration. Let me know if it's ok, but I also changed the tests to run using Go subtests so I could iterate more quickly on one test case, and added failure messages to the asserts as they were pretty cryptic before. There was one assert I simply removed because to my eyes all it did was ensure that all blocks were the default duration long - without a message I can't be sure.

File: cmd/promtool/backfill_test.go [hunk 4]
```diff
@@ -260,6 +258,46 @@ http_requests_total{code="400"} 2 1565133715.989
 		{
 			ToParse: `# HELP http_requests_total The total number of HTTP requests.
 # TYPE http_requests_total counter
+http_requests_total{code="200"} 1021 1624463088.000
+http_requests_total{code="200"} 1 1627055153.000
+http_requests_total{code="400"} 2 1627056153.000
```
Comment (codesome) [inline cmd/promtool/backfill_test.go:263]: While this is testing the max block duration, we are missing some edge cases. Here are my suggestions for the test cases:

(Reference: `ExponentialBlockRanges(2, 10, 3) -> [2 6 18 54 162 486 1458 4374 13122 39366]`)

1. 3 samples of same series such that 2 of them fall in the same block and 1 in another block. With max block duration same as right now (2048h), is 1458h.

(t, t+1400h, t+1500h)
```
http_requests_total{code="200"} 1 1624463088.000
http_requests_total{code="200"} 2 1629503088.000
http_requests_total{code="200"} 3 1629863088.000
```

2. Same as above, but with max block duration much more than `> 39366h`. So lets keep it `200000h` (basically >3x of largest block).

(t, t+39300h, t+40000h)
```
http_requests_total{code="200"} 1 1624463088.000
http_requests_total{code="200"} 2 1765943088.000
http_requests_total{code="200"} 3 1768463088.000
```

Expected number of blocks in both cases is 2.

**Comment by codesome** [comment]
> There was one assert I simply removed because to my eyes all it did was ensure that all blocks were the default duration long

Let's bring that back. We need to ensure that the blocks don't cross the expected size and also the boundaries defined by those block duration. `testBlocks` can take an additional argument for expected block duration, and the test case struct can include that info (since max is not always the block duration). To keep it simple, 0 in the struct could mean the default block duration.

**Comment by stevekuznetsov** [comment]
@codesome sure - added tests for those edges and wrote a check for the span of all blocks.

File: cmd/promtool/backfill_test.go [hunk 5]
```diff
@@ -58,12 +59,13 @@ func queryAllSeries(t testing.TB, q storage.Querier, expectedMinTime, expectedMa
...
+func testBlocks(t *testing.T, db *tsdb.DB, expectedMinTime, expectedMaxTime, expectedBlockDuration int64, expectedSamples []backfillSample, expectedNumBlocks int) {
 	blocks := db.Blocks()
-	require.Equal(t, expectedNumBlocks, len(blocks))
+	require.Equal(t, expectedNumBlocks, len(blocks), "did create correct number of blocks")
 
-	for _, block := range blocks {
-		require.Equal(t, true, block.MinTime()/tsdb.DefaultBlockDuration == (block.MaxTime()-1)/tsdb.DefaultBlockDuration)
+	for i, block := range blocks {
+		span := block.MaxTime() - block.MinTime()
+		require.Less(t, span, expectedBlockDuration, "block %d spans more time than the expected block duration", i)
```
Comment (codesome) [inline cmd/promtool/backfill_test.go:68]: It should be like before
```go
require.Equal(t, true, block.MinTime()/expectedBlockDuration == (block.MaxTime()-1)/expectedBlockDuration)
```

Reason: the block min and max times are aligned to the block duration, i.e. for a block duration 2h, a block cannot span a 2h boundary, for example `mint = 1h30m, maxt = 2h30m` is not allowed in TSDB (technically allowed, but causes issues).

Comparing after division makes sure both min and max times are in the same block bucket.

File: cmd/promtool/backfill_test.go [hunk 6]
```diff
@@ -58,12 +59,13 @@ func queryAllSeries(t testing.TB, q storage.Querier, expectedMinTime, expectedMa
 	return samples
 }
 
-func testBlocks(t *testing.T, db *tsdb.DB, expectedMinTime, expectedMaxTime int64, expectedSamples []backfillSample, expectedNumBlocks int) {
+func testBlocks(t *testing.T, db *tsdb.DB, expectedMinTime, expectedMaxTime, expectedBlockDuration int64, expectedSamples []backfillSample, expectedNumBlocks int) {
 	blocks := db.Blocks()
-	require.Equal(t, expectedNumBlocks, len(blocks))
+	require.Equal(t, expectedNumBlocks, len(blocks), "did create correct number of blocks")
```
Comment (codesome) [inline cmd/promtool/backfill_test.go:64]:
```suggestion
	require.Equal(t, expectedNumBlocks, len(blocks), "did not create correct number of blocks")
```

**Review by codesome — APPROVED** [review]
Last change and we are good to go! Thanks!

File: cmd/promtool/backfill_test.go (hunk 5 above)
Comment (codesome) [reply cmd/promtool/backfill_test.go:68]: (indirectly it also makes sure it does not span beyond the block duration)

Comment (stevekuznetsov) [reply cmd/promtool/backfill_test.go:68]: OK. I really, really do not want to expect a boolean `true` - it's impossible to read and understand when you break it. That makes sense about the alignment though.

**Comment by stevekuznetsov** [comment]
@codesome thanks for the thorough review! Hopefully the English wording in the tests makes it easier what, specifically, is being asserted about for the next bloke to break them ;)

**Comment by codesome** [comment]
Closed and reopened to re-run the CI as they were stuck

**Comment by codesome** [comment]
CI is failing, looks like you are missing gofmt on your changes :)

**Comment by stevekuznetsov** [comment]
@codesome haha sorry! Fighting with a new editor setup :facepalm:

**Comment by stevekuznetsov** [comment]
@codesome i'm not sure the CircleCI test is running my latest code - you can see in the output that two blocks are written, but the `tsdb` read after only sees one. I saw this before I increased the retention duration in the tests. Locally, using my current code, I cannot reproduce the test failure:

```
$ go test ./cmd/promtool/... -run TestBackfill/Maximum_block_duration_longer_than_longest_possible_duration,_uses_largest_duration,_puts_all_data_into_two_blocks. -count 100
ok  	github.com/prometheus/prometheus/cmd/promtool	239.161s
```

How can we re-run the test... ?

**Comment by stevekuznetsov** [comment]
@codesome sorry about that ... looks all green now :)
